load all of the user’s chosen font files as fallbacks for ImGui to use

db48x opened the pull request:
#### Summary
Interface "load all of the user’s chosen font files as fallbacks for ImGui to use"

#### Purpose of change
This allows imgui to pull any glyphs missing from the user’s first choice of font from the others in the list.

#### Testing
Set the game locale to Chinese or Japanese. Add an appropriate font to the `gui_font` list in `fonts.json`. Use the ImGui style editor to verify that there are glyphs loaded for both CJK characters and the usual Latin characters.

#### Additional context
I had to wait until ocornut/imgui#8081 was fixed before I could implement this. Now that it’s been fixed and we’ve caught up to a version of ImGui that includes the fix there’s no reason to avoid it.

db48x committed:
load all of the user’s chosen font files as fallbacks for imgui to use

If a glyph isn’t found in the user’s first choice it’ll be taken from
one of the others. This is especially useful for CJK languages where
the font might not actually include any latin characters.

CLIDragon commented:
Can you confirm whether the loading issue I mentioned in https://github.com/CleverRaven/Cataclysm-DDA/pull/78978#issuecomment-2572241898 still occurs with this merged?

db48x commented:
Funny you should ask! After testing, I find that it suffers from the same bug. The fonts are loaded, but the atlas is a solid color.

GuardianDll commented:
Is it ready to merge?

db48x commented:
Technically it can be merged. It won't cause any additional bugs :)

However I would like to figure out what causes a missing font to break ImGui’s font atlas. It is something that `debugmsg` is doing when it displays the error message; that much is obvious.

db48x committed:
warn about missing font files via printf, not debugmsg

debugmsg triggers a new ImGui frame, which messes up the font
atlas. printf is suboptimal, but I plan to introduce an in–game font
picker that can inform the user that the font file couldn’t be found
in a less heavy–handed and totalizing way than debugmsg. Therefore a
suppose printf will do for now.

db48x commented:
oops
